CMakeLists.txt

```diff
@@ -208,298 +229,321 @@ set(AVR_SOURCES
-
-    target_compile_definitions(${FW_EN} PUBLIC LANG_MODE=0)
+  # Set FW_SOURCES to be reproducible in this variant as it's set in a separate project
+  set_reproducible_sources(FW_SOURCES "Firmware/")
+
+  # Create the Configuration_Prusa.h for this variant so it can be #included.
+  set(VARIANT_CFG_DIR ${CMAKE_CURRENT_BINARY_DIR}/include)
+  set(VARIANT_CFG_DST ${VARIANT_CFG_DIR}/Configuration_prusa.h)
+  set(VARIANT_CFG_SRC ${CMAKE_SOURCE_DIR}/Firmware/variants/${variant_name}.h)
+  add_custom_command(
+    OUTPUT ${VARIANT_CFG_DST}
+    COMMAND ${CMAKE_COMMAND} -E copy ${VARIANT_CFG_SRC} ${VARIANT_CFG_DST}
+    COMMENT "Generating Configuration_prusa.h for ${variant_name}"
+    BYPRODUCTS ${VARIANT_CFG_DIR}
+    DEPENDS ${VARIANT_CFG_SRC}
+    )
+  string(REPLACE "1_75mm_" "" variant_name "${variant_name}")
+  string(REPLACE "-E3Dv6full" "" variant_name "${variant_name}")
+
+  # Single-language build
+  set(FW_EN "${variant_name}_EN-only")
+  set(FW_HEX ${CMAKE_BINARY_DIR}/${FN_PREFIX}-${FW_EN}.hex)
+
+  add_base_binary(${FW_EN})
+  target_compile_definitions(${FW_EN} PUBLIC LANG_MODE=0)
+  add_custom_command(
+    TARGET ${FW_EN}
+    POST_BUILD
+    COMMAND ${CMAKE_OBJCOPY} -O ihex ${FW_EN} ${FW_EN}.hex
+    COMMAND ${CMAKE_OBJCOPY} -O ihex ${FW_EN} ${FW_HEX}
+    BYPRODUCTS ${FW_EN}.hex ${FW_HEX}
+    COMMENT "Generating ${FW_EN}.hex"
+    )
+  add_dependencies(ALL_ENGLISH ${FW_EN})
+
+  # Multi-language build/s
+  set(FW_LANG_BASE "${variant_name}_Multilang_base")
+  set(FW_LANG_PATCH "${variant_name}_Multilang_patch")
+  add_base_binary(${FW_LANG_BASE})
+  target_compile_definitions(${FW_LANG_BASE} PUBLIC LANG_MODE=1)
+
+  # Construct language map
+  set(LANG_TMP_DIR lang)
+  set(LANG_MAP ${LANG_TMP_DIR}/${variant_name}_lang.map)
+
+  add_custom_command(
+    OUTPUT ${LANG_MAP}
+    COMMAND ${CMAKE_OBJCOPY} -O binary ${FW_LANG_BASE} ${FW_LANG_PATCH}.bin
+    COMMAND ${CMAKE_SOURCE_DIR}/lang/lang-map.py ${FW_LANG_BASE} ${FW_LANG_PATCH}.bin > ${LANG_MAP}
+    COMMAND ${CMAKE_OBJCOPY} -I binary -O ihex ${FW_LANG_PATCH}.bin ${FW_LANG_PATCH}.hex
+    DEPENDS ${FW_LANG_BASE}
+    BYPRODUCTS ${FW_LANG_PATCH}.bin ${FW_LANG_PATCH}.hex
+    COMMENT "Generating ${variant_name} language map"
+    )
+
+  # Build language catalogs
+  set(LANG_BINS "")
+  foreach(LANG IN LISTS SELECTED_LANGUAGES)
+    set(LANG_BIN ${LANG_TMP_DIR}/${variant_name}_${LANG}.bin)
+    set(PO_FILE "${CMAKE_SOURCE_DIR}/lang/po/Firmware_${LANG}.po")
+
     add_custom_command(
-        TARGET ${FW_EN}
-        POST_BUILD
-        COMMAND ${CMAKE_OBJCOPY} -O ihex ${CMAKE_CURRENT_BINARY_DIR}/${FW_EN} ${CMAKE_BINARY_DIR}/${FN_PREFIX}-${FW_EN}.hex
-        BYPRODUCTS ${CMAKE_BINARY_DIR}/${FN_PREFIX}-${FW_EN}.hex
-        COMMENT "Generating ${variant_name} hex"
+      OUTPUT ${LANG_BIN}
+      #[[
+      # Check po file:
+      #COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/lang/lang-check.py --no-warning --map ${LANG_MAP} ${PO_FILE}
+      #]]
```

Comment:
This should probably be re-enabled for the language builds... no?
I turned it off because it wasn't working for me but it is used by PF-build.

Reply:
Yup. But this (and the following) were commented before and remained commented as I didn't have time to check them.

In this case, lang-check will emit huge amounts of output with the current build layout due to the number of targets. IMHO it needs some sort of special handling.